packages/application-components/src/components/portals-container/portals-container.tsx

```diff
@@ -79,48 +114,110 @@ const useObserverElementDimensions = (
+                ${props.containerSelectorToPreventScrollingOnOpen} {
+                overflow: hidden;
+              }
 
-            .ReactModal__Body--open #${PORTALS_CONTAINER_ID} {
-              position: fixed;
-              height: calc(
-                100% - ${props.offsetTop} -
-                  ${globalNotificationsElementDimensions.height}px -
-                  ${pageNotificationsElementDimensions.height}px
-              );
-              width: calc(100% - ${props.offsetLeft});
-              top: calc(
-                ${props.offsetTop} +
-                  ${globalNotificationsElementDimensions.height}px +
-                  ${pageNotificationsElementDimensions.height}px
-              );
-              right: 0;
-              bottom: 0;
-              z-index: ${props.zIndex};
-            }
+              .ReactModal__Body--open #${PORTALS_CONTAINER_ID} {
+                position: fixed;
+                height: calc(
+                  100% - ${props.offsetTop} -
+                    ${globalNotificationsElementDimensions.height}px -
+                    ${pageNotificationsElementDimensions.height}px
+                );
+                width: calc(100% - ${props.offsetLeft});
+                top: calc(
+                  ${props.offsetTop} +
+                    ${globalNotificationsElementDimensions.height}px +
+                    ${pageNotificationsElementDimensions.height}px
+                );
+                right: 0;
+                bottom: 0;
+                z-index: ${props.zIndex};
+              }
 
-            .ReactModal__Body--open.body__menu-open #${PORTALS_CONTAINER_ID} {
-              width: calc(100% - ${props.offsetLeftOnExpandedMenu});
-            }
-          `}
+              .ReactModal__Body--open.body__menu-open #${PORTALS_CONTAINER_ID} {
+                width: calc(100% - ${props.offsetLeftOnExpandedMenu});
+              }
+            `,
+            // Apply styles for stacking layers.
+            ...stackingLayers.map(
+              /**
+               * Every "overlay" container gets assigned a sequential `z-index` level.
+               * Furthermore, the "modal overlay" containers (not "dialog") get assigned
+               * the correct indentation level width.
+               */
+              (stackingLayer) => css`
+                .ReactModalPortal[data-level='${stackingLayer.stackingLevel}']
+                  [data-role$='overlay'] {
+                  z-index: calc(
+                    ${props.baseModalZIndex} + ${stackingLayer.stackingLevel}
+                  );
+                }
+                .ReactModalPortal[data-level='${stackingLayer.stackingLevel}']
+                  [data-role='modal-overlay']
+                  [role='dialog'] {
+                  width: calc(
+                    100% -
+                      (${indentationSize} * ${stackingLayer.indentationLevel})
+                  );
+                }
+              `
```

Comment:
Something that confused me is that we are creating and so updating some global css rules in the document every time we add/remove a new `overlay` to then link them to the actual HTML elements with a class.
My first impression was that this could be done inline. Instead of adding a new data attribute to the overlay elements, just add to them the `z-index` and `width` css rules inline. Perhaps with that approach we could also remove the local state since we are resetting it on every element added to the `PortalsContainer`.

Maybe I could try to explain myself better when if have a talk about this PR.

Reply:
Yes I know what you mean. Initially I tried to do that but ended up here instead.

I'll give it another try, thanks for suggesting that!

Reply:
Ok I double checked and then I remembered the reason why I used CSS.

There are cases were modal components are rendered but they are not "opened/active" yet, for example a confirmation dialog is rendered on the page but it hasn't been opened yet.

However, the portal container of these components is rendered into the DOM even though it's empty. Then when the dialog is opened, it renders inside its portal container.

![image](https://user-images.githubusercontent.com/1110551/166554479-6a3ce4f5-e08d-40e2-a8fd-5fa2f40a000c.png)


Because the container was already rendered, the mutation observer won't be notified again when the dialog actually opens. In theory we could also subscribe to the `subtree` changes but it will trigger a mutation for the entire subtree of nodes.

So by using pre-defined CSS selectors, when the dialog opens, the styles gets automaticaly applied.

Reply:
Ah! I see.

Thanks for the explanation.